web/ui/react-app/src/ExpressionInput.tsx

```diff
@@ -0,0 +1,159 @@
+import $ from 'jquery';
+import React, { Component } from 'react';
+import {
+  Button,
+  InputGroup,
+  InputGroupAddon,
+  InputGroupText,
+  Input,
+} from 'reactstrap';
+
+import Downshift, { ChildrenFunction, ControllerStateAndHelpers, DownshiftInterface } from 'downshift';
+import fuzzy from 'fuzzy';
+
+import { library } from '@fortawesome/fontawesome-svg-core';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import { faSearch, faSpinner } from '@fortawesome/free-solid-svg-icons';
+
+library.add(faSearch, faSpinner);
+
+interface ExpressionInputProps {
+  value: string;
+  metricNames: string[];
+  executeQuery: (expr: string) => void;
+  loading: boolean;
+}
+
+class ExpressionInput extends Component<ExpressionInputProps> {
+  prevNoMatchValue: string | null = null;
+  private exprInputRef = React.createRef<HTMLInputElement>();
+
+  handleKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'Enter' && !event.shiftKey) {
+      this.props.executeQuery(this.exprInputRef.current!.value);
+      event.preventDefault();
+    }
+  }
+
+  renderAutosuggest = (downshift: any) => {
+    if (!downshift.isOpen) {
+      return null;
+    }
+
+    if (this.prevNoMatchValue && downshift.inputValue.includes(this.prevNoMatchValue)) {
+      return null;
+    }
+
+    let matches = fuzzy.filter(downshift.inputValue.replace(/ /g, ''), this.props.metricNames, {
+      pre: "<strong>",
+      post: "</strong>",
+    });
+
+    if (matches.length === 0) {
+      this.prevNoMatchValue = downshift.inputValue;
+      return null;
+    }
+
+    return (
+      <ul className="autosuggest-dropdown" {...downshift.getMenuProps()}>
+        {
+          matches
+            .slice(0, 200) // Limit DOM rendering to 100 results, as DOM rendering is sloooow.
+            .map((item, index) => (
+              <li
+                {...downshift.getItemProps({
+                  key: item.original,
+                  index,
+                  item: item.original,
+                  style: {
+                    backgroundColor:
+                      downshift.highlightedIndex === index ? 'lightgray' : 'white',
+                    fontWeight: downshift.selectedItem === item ? 'bold' : 'normal',
+                  },
+                })}
+              >
+                {/* TODO: Find better way than setting inner HTML dangerously. We just want the <strong> to not be escaped.
+                    This will be a problem when we save history and the user enters HTML into a query. */}
+                <span dangerouslySetInnerHTML={{__html: item.string}}></span>
```

Comment:
To address this concern [sanitize-html](https://www.npmjs.com/package/sanitize-html) library can be used which allows whitelisting certain HTML elements to be allowed.

Reply:
Nice! Would you be up for doing a PR after this one? The only thing we need to watch out for is that it still performs well for thousands of list entries, but I hope that will be fine.

At the moment, there is no query history search in the new UI yet, so it's not an immediate vulnerability.

Reply:
Yes, I'd love to.

Reply:
🙌 Can't wait for people who actually know frontend / React stuff to contribute and make this all better :)

Reply:
Created PR which addresses this - https://github.com/prometheus/prometheus/pull/6165/files